File: experiments-local/experiment-15/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep2/src/validation.ts
import type { BookInput, ValidationError } from "./types.ts";

/**
 * Validate book input for create/update operations.
 * Returns a list of validation errors (empty if valid).
 *
 * For updates (partial=true), only fields that are present are validated.
 * For creates (partial=false), title and author are required.
 */
export function validateBook(input: BookInput, partial = false): ValidationError[] {
  const errors: ValidationError[] = [];

  const hasTitle = input.title !== undefined;
  const hasAuthor = input.author !== undefined;

  if (!partial || hasTitle) {
    if (!hasTitle || typeof input.title !== "string" || input.title.trim() === "") {
      errors.push({ field: "title", message: "title is required and must be a non-empty string" });
    }
  }

  if (!partial || hasAuthor) {
    if (!hasAuthor || typeof input.author !== "string" || input.author.trim() === "") {
      errors.push({ field: "author", message: "author is required and must be a non-empty string" });
    }
  }

  if (input.year !== undefined && input.year !== null) {
    if (typeof input.year !== "number" || !Number.isInteger(input.year) || input.year < 0 || input.year > 9999) {
      errors.push({ field: "year", message: "year must be an integer between 0 and 9999" });
    }
  }

  if (input.isbn !== undefined && input.isbn !== null) {
    if (typeof input.isbn !== "string" || input.isbn.trim() === "") {
      errors.push({ field: "isbn", message: "isbn must be a non-empty string or null" });
    }
  }

  return errors;
}

File: experiments-local/experiment-opencode-easy-final/runs/agent=opencode_language=go_model=openrouter/z-ai/glm-5.2_tooling=none/rep1/validation.go
package main

import "errors"

var ErrNotFound = errors.New("book not found")

type ValidationError struct {
	Fields map[string]string
}

func (e *ValidationError) Error() string {
	msg := "validation error"
	for k, v := range e.Fields {
		msg += "; " + k + ": " + v
	}
	return msg
}

func validateBook(in bookInput) (*Book, error) {
	verr := &ValidationError{Fields: map[string]string{}}
	b := &Book{}

	if in.Title == nil || *in.Title == "" {
		verr.Fields["title"] = "title is required"
	} else {
		b.Title = *in.Title
	}

	if in.Author == nil || *in.Author == "" {
		verr.Fields["author"] = "author is required"
	} else {
		b.Author = *in.Author
	}

	if in.Year != nil {
		b.Year = *in.Year
	}
	if in.ISBN != nil {
		b.ISBN = *in.ISBN
	}

	if len(verr.Fields) > 0 {
		return nil, verr
	}
	return b, nil
}

func validatePartial(in bookInput) (*Book, error) {
	verr := &ValidationError{Fields: map[string]string{}}
	b := &Book{}

	if in.Title == nil || *in.Title == "" {
		verr.Fields["title"] = "title is required"
	} else {
		b.Title = *in.Title
	}
	if in.Author == nil || *in.Author == "" {
		verr.Fields["author"] = "author is required"
	} else {
		b.Author = *in.Author
	}
	if in.Year != nil {
		b.Year = *in.Year
	}
	if in.ISBN != nil {
		b.ISBN = *in.ISBN
	}

	if len(verr.Fields) > 0 {
		return nil, verr
	}
	return b, nil
}
